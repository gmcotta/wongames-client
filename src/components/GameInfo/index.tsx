import { FavoriteBorder } from '@styled-icons/material-outlined'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from 'components/Button'
import Heading from 'components/Heading'
import Ribbon from 'components/Ribbon'
import * as S from './styles'

export type GameInfoProps = {
  title: string
  description: string
  price: string
}

const GameInfo = ({ title, description, price }: GameInfoProps) => (
  <S.Wrapper>
    <Heading lineBottom color="black">
      {title}
    </Heading>
    <S.Description>{description}</S.Description>
    <Ribbon color="secondary">{price}</Ribbon>
    <S.ButtonWrapper>
      <Button size="large" icon={<AddShoppingCart />}>
        Add to cart
      </Button>
      <Button size="large" minimal icon={<FavoriteBorder />}>
        Wish list
      </Button>
    </S.ButtonWrapper>
  </S.Wrapper>
)

export default GameInfo
