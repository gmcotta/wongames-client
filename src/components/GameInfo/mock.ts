import { GameInfoProps } from '.'

const mock: GameInfoProps = {
  title: 'Borderland 3',
  description:
    'Agora é a hora de eliminar Handsome Jack. Reúna seus amigos para correr desenfreadamente pelo mundo coletando milhões de itens.',
  price: 215
}

export default mock
