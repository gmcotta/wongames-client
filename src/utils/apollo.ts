import { useMemo } from 'react'
import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
  HttpLink
} from '@apollo/client'

let apolloClient: ApolloClient<NormalizedCacheObject>

function createApolloClient() {
  return new ApolloClient({
    link: new HttpLink({ uri: 'http://localhost:1337/graphql' }),
    cache: new InMemoryCache(),
    ssrMode: typeof window === 'undefined'
  })
}

export function initializeApollo(initialState = {}) {
  const apolloClientGlobal = apolloClient ?? createApolloClient()
  if (initialState) {
    apolloClientGlobal.cache.restore(initialState)
  }
  if (typeof window === 'undefined') return apolloClientGlobal
  apolloClient = apolloClient ?? apolloClientGlobal
  return apolloClient
}

export function useApollo(initialState = {}) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
