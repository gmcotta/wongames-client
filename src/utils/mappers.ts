import { QueryGames_games } from 'graphql/generated/QueryGames'
import {
  QueryHome_banners,
  QueryHome_sections_freeGames_highlight
} from 'graphql/generated/QueryHome'

export function bannerMapper(banners: QueryHome_banners[]) {
  return banners.map((banner) => ({
    img: `http://localhost:1337${banner.image?.url}`,
    title: banner.title,
    subtitle: banner.subtitle,
    buttonLabel: banner.button?.label,
    buttonLink: banner.button?.link,
    ...(banner.ribbon && {
      ribbonText: banner.ribbon.text,
      ribbonSize: banner.ribbon.size,
      ribbonColor: banner.ribbon.color
    })
  }))
}

export function gamesMapper(games: QueryGames_games[] | null | undefined) {
  return games?.map((game) => ({
    title: game.name,
    slug: game.slug,
    developer: game.developers[0].name,
    img: `http://localhost:1337${game.cover?.url}`,
    price: game.price
  }))
}

export function highlightMapper(
  highlight: QueryHome_sections_freeGames_highlight | null | undefined
) {
  return (
    highlight && {
      title: highlight.title,
      subtitle: highlight.subtitle,
      buttonLabel: highlight.buttonLabel,
      buttonLink: highlight.buttonLink,
      backgroundImage: `http://localhost:1337${highlight.background?.url}`,
      floatImage: `http://localhost:1337${highlight.floatImage?.url}`,
      contentAlignment: highlight.contentAlignment
    }
  )
}
