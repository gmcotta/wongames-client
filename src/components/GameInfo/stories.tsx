import { Story, Meta } from '@storybook/react'
import GameInfo, { GameInfoProps } from '.'

export default {
  title: 'GameInfo',
  component: GameInfo
} as Meta

export const Default: Story<GameInfoProps> = (args) => <GameInfo {...args} />

Default.args = {
  title: 'Borderland 3',
  description:
    'Agora é a hora de eliminar Handsome Jack. Reúna seus amigos para correr desenfreadamente pelo mundo coletando milhões de itens.',
  price: '215.00'
}
