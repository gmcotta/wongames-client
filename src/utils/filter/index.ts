import { ItemProps } from 'components/ExploreSidebar'
import { ParsedUrlQueryInput } from 'querystring'

export type FilterItemProps = Pick<ItemProps, 'name' | 'type'>

type ParseArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: FilterItemProps[]
}

export const parseQueryStringToWhere = ({
  queryString,
  filterItems
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}
  Object.keys(queryString)
    .filter((item) => item !== 'sort')
    .forEach((key) => {
      const item = filterItems?.find((item) => item.name === key)
      const isNotCheckbox = item?.type !== 'checkbox'
      obj[key] = isNotCheckbox
        ? queryString[key]
        : { name_contains: queryString[key] }
    })
  return obj
}

export const parseQueryStringToFilter = ({
  queryString,
  filterItems
}: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}
  Object.keys(queryString).forEach((key) => {
    const item = filterItems.find((item) => item.name === key)
    const isCheckbox = item?.type === 'checkbox'
    const isArray = Array.isArray(queryString[key])
    obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key]
  })
  return obj
}
