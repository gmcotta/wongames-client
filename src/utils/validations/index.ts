import joi from 'joi'
import { SignInValues } from 'components/SignInForm'
import { UsersPermissionsRegisterInput } from 'graphql/generated/globalTypes'

const fieldValidations = {
  username: joi.string().min(5).required(),
  email: joi
    .string()
    .email({ tlds: { allow: false } })
    .required(),
  password: joi.string().required(),
  confirmPassword: joi.string().valid(joi.ref('password')).required().messages({
    'any.only': 'confirm password does not match with password'
  })
}
export type FieldErrors = {
  [key: string]: string
}
function getFieldErrors(objError: joi.ValidationResult) {
  const errors: FieldErrors = {}
  if (objError.error) {
    objError.error.details.forEach((error) => {
      errors[error.path.join('.')] = error.message
    })
  }
  return errors
}

export function signInValidate(values: SignInValues) {
  const { email, password } = fieldValidations
  const schema = joi.object({ email, password })
  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

export function signUpValidate(values: UsersPermissionsRegisterInput) {
  const schema = joi.object(fieldValidations)
  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
