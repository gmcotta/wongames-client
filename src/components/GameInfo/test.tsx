import { screen, render } from 'utils/testUtils'

import GameInfo from '.'

const props = {
  id: '1',
  title: 'Borderland 3',
  description: 'game description',
  price: 210
}

describe('<GameInfo />', () => {
  it('should render the game info', () => {
    const { container } = render(<GameInfo {...props} />)
    expect(
      screen.getByRole('heading', { name: /borderland 3/i })
    ).toBeInTheDocument()
    expect(screen.getByText(/game description/i)).toBeInTheDocument()
    expect(screen.getByText(/210.00/i)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render buttons', () => {
    render(<GameInfo {...props} />)
    expect(
      screen.getByRole('button', { name: /add to cart/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('button', { name: /wish list/i })
    ).toBeInTheDocument()
  })
})
