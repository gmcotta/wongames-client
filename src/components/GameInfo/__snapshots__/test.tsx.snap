// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameInfo /> should render the game info 1`] = `
.c8 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  cursor: pointer;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 5rem;
  font-size: 1.6rem;
  padding: 1.2rem 5.2rem;
}

.c7 svg {
  width: 1.5rem;
}

.c7 svg + span {
  margin-left: 0.8rem;
}

.c7:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  cursor: pointer;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 5rem;
  font-size: 1.6rem;
  padding: 1.2rem 5.2rem;
  background: none;
  color: #F231A5;
}

.c9 svg {
  width: 1.5rem;
}

.c9 svg + span {
  margin-left: 0.8rem;
}

.c9:hover {
  background: none;
  -webkit-filter: brightness(0.9);
  filter: brightness(0.9);
}

.c2 {
  color: #030517;
  position: relative;
  margin-bottom: 3.2rem;
  font-size: 2.0rem;
}

.c2::after {
  content: '';
  position: absolute;
  left: 0;
  bottom: -1rem;
  width: 5rem;
  border-bottom: 0.5rem solid #F231A5;
}

.c5 {
  background-color: #3CD3C1;
  right: -2rem;
  height: 3.6rem;
  font-size: 1.4rem;
  padding: 0.6rem 4rem;
  position: absolute;
  top: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  font-weight: 600;
}

.c5::after {
  color: #3CD3C1;
  -webkit-filter: brightness(0.44);
  filter: brightness(0.44);
}

.c5::after {
  top: 3.6rem;
  border-top-width: 1rem;
  border-right-width: 2rem;
}

.c5::after {
  content: '';
  position: absolute;
  right: 0;
  border-style: solid;
  border-left-width: 0rem;
  border-right-color: transparent;
  border-bottom-color: transparent;
  border-bottom-width: 1rem;
}

.c0 {
  position: relative;
  background-color: #FAFAFA;
  padding: 2.4rem;
}

.c0 .c1 {
  max-width: 72%;
}

.c0 .c4 {
  right: -1rem;
}

.c0 .c4::after {
  border-right-width: 1rem;
}

.c3 {
  font-size: 1.4rem;
  color: #8F8F8F;
  margin-bottom: 2.4rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 > button {
  width: 100%;
  margin-bottom: 0.8rem;
}

@media (min-width:768px) {
  .c2 {
    font-size: 2.8rem;
  }
}

@media (min-width:768px) {
  .c0 .c1 {
    max-width: 78%;
  }
}

@media (min-width:768px) {
  .c0 .c4 {
    right: 2.4rem;
    top: 2.4rem;
    font-size: 1.8rem;
  }

  .c0 .c4::after {
    border: none;
  }
}

@media (min-width:768px) {
  .c3 {
    max-width: 77rem;
  }
}

@media (min-width:768px) {
  .c6 {
    -webkit-flex-direction: row-reverse;
    -ms-flex-direction: row-reverse;
    flex-direction: row-reverse;
  }

  .c6 > button {
    width: initial;
    margin-bottom: 0;
  }
}

<div
  class="c0"
>
  <h2
    class="c1 c2"
    color="black"
  >
    Borderland 3
  </h2>
  <p
    class="c3"
  >
    game description
  </p>
  <div
    class="c4 c5"
    color="secondary"
  >
    R$ 
    210.00
  </div>
  <div
    class="c6"
  >
    <button
      class="c7"
    >
      <svg
        aria-hidden="true"
        class="c8"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-8.9-5h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4l-3.87 7H8.53L4.27 2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2z"
        />
      </svg>
      <span>
        Add to cart
      </span>
    </button>
    <button
      class="c9"
    >
      <svg
        aria-hidden="true"
        class="c8"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
      <span>
        Wish list
      </span>
    </button>
  </div>
</div>
`
